pkg/beacon/beacon.go

```diff
@@ -175,8 +175,58 @@ func Initialize(
 				onGroupSelected,
 			)
 			if err != nil {
-				logger.Errorf("Tickets submission failed: [%v]", err)
+				logger.Errorf("tickets submission failed: [%v]", err)
 			}
+
+			// We deduplicate `GroupSelectionStarted` events by keeping them
+			// into the `pendingGroupSelections` cache. Duplicated events
+			// may occur due to chain reorgs or because of the subscription
+			// monitoring mechanism which re-fetches past events in case
+			// something bad happens with the websocket connection pointing
+			// to the Ethereum node. However, we can't remove current group
+			// selection from cache just after ticket submission terminates
+			// because a duplicated event coming right after the removal
+			// will trigger a new DKG which will break the current one.
+			// We need to hold the cache entry for the maximum DKG duration
+			// time. We should also do it regardless of the outcome on this
+			// node as other nodes may still perform DKG successfully.
+			// Removing the cache entry after hitting the DKG timeout is
+			// enough and we don't need to monitor for a successful group
+			// registration. A successful group registration means we never
+			// have to trigger group selection again for the group seed we hold
+			// in the cache so we don't have to bother about removing the cache
+			// entry just after the group is registered. In contrary, when
+			// current group selection fails, we need to remove the cache
+			// entry once the timeout is hit to allow the node to retry
+			// the group selection for that seed as soon as possible.
+
+			dkgDurationBlocks :=
+				chainConfig.TicketSubmissionTimeout +
+					gjkr.ProtocolBlocks() +
+					dkgresult.PrePublicationBlocks() +
+					(uint64(chainConfig.GroupSize) * chainConfig.ResultPublicationBlockStep)
+
+			dkgTimeout := event.BlockNumber + dkgDurationBlocks
+
+			if err := blockCounter.WaitForBlockHeight(dkgTimeout); err != nil {
+				// We assume the average block time is 15s.
+				delayDuration := time.Duration(dkgDurationBlocks*15) * time.Second
+
+				logger.Errorf(
+					"could not set the block counter to wait "+
+						"until the DKG timeout block [%v] to remove the "+
+						"current group selection from cache: [%v]; "+
+						"current group selection will be removed from cache "+
+						"after a fixed delay of [%v]",
+					dkgTimeout,
+					err,
+					delayDuration,
+				)
+
+				<-time.After(delayDuration)
```

Comment:
I'd rather emit an error from the deduplicator and not allow the client to start ticket submission than to assume the block time. I think it's less error-prone and if one does not submit their ticket nothing bad happens.

Reply:
This is not so simple. If we emit an error here we must remove the current group selection seed from `GroupSelectionTrack` deduplicator immediately. That means we can still obtain a duplicated `GroupSelectionStarted` event and the client will begin a new ticket submission process. That round will fail as the contract state is "we are after ticket submission, DKG is in progress" but at the end, `SelectedParticipants` will return a valid participant list (from previous legit ticket submission) and the node will start a new DKG on behalf of its virtual stakers. This new DKG will break the ongoing one executed by other nodes as we are observing on `keep-test`. This is why we must wait with cache removal for a certain period of time. This period should ideally be the moment when DKG timeout occurs as we are sure the recent DKG is completed. However, in case the block counter returns an error (rather unlikely) we must wait a period of time close enough to the DKG duration. It doesn't have to be an exact value and I think the presented way is fair enough. Unfortunately, there is no ideal solution here. But, even if real block time is bigger than 15s, and the cache is cleared a little bit too early it will be not so easy to break the ongoing DKG at this stage. We still need a duplicated `GroupSelectionStarted` event which is hard to occur after such a long time from the original event. Also, even if it occurs, the original DKG should be close to their end and a new duplicated DKG will have a hard work to break things as members from the original DKG will not accept messages from members of duplicated DKG.

Reply:
I see. In this case, maybe it makes more sense to store last group selection start event block number instead of storing the group selection seed in the event track map? We would allow another group selection to start if the last group selections start block + DKG time is less than the current block height.

In this case, we do not have to delete anything, we are just storing one integer.

Reply:
Good idea! Done in: https://github.com/keep-network/keep-core/pull/2574/commits/ea353ddbeff0915b5e9a11c030515c461707ff35

Reply:
But, we should keep in mind that if one DKG terminates with success, the next one will not be processed until DKG execution time elapses from the start of the latest DKG.

Reply:
I like where it is heading but I think we could simplify and optimize the logic. Also, I think it makes sense to solve the problem of the relay request event deduplication in this PR so that we do not have to go back to it after a couple of days. After reading your TODO I realized it's enough to have two relay requests executed one after another to break the deduplication logic.

There are two problems I see:
- Relay request deduplication does not work as intended; It deduplicates events only for a very short period of time.
- DKG deduplication is not efficient; We execute on-chain checks for most of the events that are getting deduplicated instead of deduplicating most of them using in-memory track first, and reaching to the chain only when it is absolutely necessary.

What I'd suggest doing for relay request:
Let's get rid of `NotifyRelayEntryCompleted`. As you noticed in the TODO, all the logic is done asynchronously and we currently can't provide a clear "completed" signal. We might want to change it in the future but without significant refactoring, it is not possible. Let's pass to the deduplicator how long it takes to execute relay entry signing protocol *without* the part for the result publication. When notifying about signing started, let's pass both block number, group public key, and previous entry (or the whole event). In case the block number is < than the last event block number from the track plus entry signing protocol execution time, the previous entry is the same, and group public key is the same, we consider the event as a duplicate given that no new signing could happen in that time for the same entry and group. If the block number is not under the threshold, we need to check the chain if signing is in progress and what is the previous relay entry. So two on-chain calls, that could be in future reduced to one (first we need to clean up operator contract).

What I'd suggest doing for DKG:
Let's pass the time it takes to select a group and execute DKG *without* a time to submit the result to the deduplicator. If we see an event under the threshold of the last DKG start block plus that time, we consider it as a duplicate. If the start block is above the threshold, we let it go.

Reply:
Ok, you're right and we can simplify as you proposed. However, for relay entry, we can't determine the duration in blocks as we have in DKG because the relay entry protocol is time-based. We also don't need to check the relay entry target group as requests occur synchronously. Here is my proposed solution: https://github.com/keep-network/keep-core/pull/2574/commits/4f0ac71660f7153909d79da5c874cbde38bfcf58